resolver/addressable_resolver.go

```diff
@@ -184,30 +247,24 @@ func (r *URIResolver) URIFromObjectReference(ctx context.Context, ref *corev1.Ob
 		return nil, fmt.Errorf("failed to get object %s/%s: %w", ref.Namespace, ref.Name, err)
 	}
 
-	// K8s Services are special cased. They can be called, even though they do not satisfy the
-	// Callable interface.
-	if ref.APIVersion == "v1" && ref.Kind == "Service" {
-		url := &apis.URL{
-			Scheme: "http",
-			Host:   network.GetServiceHostname(ref.Name, ref.Namespace),
-			Path:   "",
-		}
-		return url, nil
-	}
-
 	addressable, ok := obj.(*duckv1.AddressableType)
 	if !ok {
 		return nil, apierrs.NewBadRequest(fmt.Sprintf("%+v (%T) is not an AddressableType", ref, ref))
 	}
+	if addressable.Status.Addresses != nil && len(addressable.Status.Addresses) != 0 {
+		if addressable.Status.Address == nil {
+			return &addressable.Status.Addresses[0], nil
+		} else {
+			for _, addr := range addressable.Status.Addresses {
+				if addr.Name == addressable.Status.Address.Name {
+					return &addr, nil
+				}
+			}
+			return nil, apierrs.NewBadRequest(fmt.Sprintf("address with name %s cannot be found/resolved for %+v", addressable.Status.Address.Name, ref))
+		}
+	}
 	if addressable.Status.Address == nil {
 		return nil, apierrs.NewBadRequest(fmt.Sprintf("address not set for %+v", ref))
 	}
-	url := addressable.Status.Address.URL
-	if url == nil {
-		return nil, apierrs.NewBadRequest(fmt.Sprintf("URL missing in address of %+v", ref))
-	}
-	if url.Host == "" {
-		return nil, apierrs.NewBadRequest(fmt.Sprintf("hostname missing in address of %+v", ref))
-	}
```

Comment:
Have you considered leaving these checks as they were? They seem useful to report back a comprehensive error message